### ISSUE TITLE
Storage: 403 Forbidden, mishandled(?) for parallel processes (intrusion alerts)

### ISSUE THREAD

**Issue by ThatGuyFromFinland** [GoogleCloudPlatform/google-cloud-go#461]
I received a request for appeal, due to violating usage policies by activity, which looks like an intrusion attempt against a third-party. We are using Web-crawlers, so my first guess was a misconfigured wget; however, our crawlers have been using same parameters for two months now.

Then I noticed, that shortly before the request for appeal, one of our crawlers had been launched with wrong credentials. I believe the correct behavior for the SDK would be to only retry using same object handler if credentials have changed, instead of retrying the API with each subsequent write request after receiving 403.

Here is what happened:
-Our crawler was started with improper credential configuration
-We crawl domains and then stream the content to Google Cloud Storage using the Go SDK in a "massively" parallel process
-In case you attempt to use the SDK with bad credentials, e.g. wrong region, it will look like brute-force attack.

Of course I could write my own wrapper, but that will not prevent any future intrusion alerts you will receive when other developers launch their parallel process with wrong credentials, but forget to check for the 403 for a specific object. Such wrapper is not hard to write in my case, but in some cases (I believe) there might be problems when locking your API calls immmediately for all running processes might be more difficult; especially with SDKs for languages, that are less designed for parallelism than Go (trying to write a wrapper without architecture, that uses channels would be more challenging).

Of course, when using Kubernetes or similar, this wouldn't fix the problem entirely.

I believe you understand what I mean, and can make the best decision based on this information. Even though it is not the correct way, to not use service credentials, we developers at sometimes want to try the easy way out and forget, that it will blow up everything. IMO good SDK doesn't let you learn terrible things, like mistakenly done brute-force intrusion alerts, the hard-way; good SDK is gentle and forgiving :)

This also adds complexity for us, if we will expand our GCP-team, because the new guys should face a gentle and forgiving environment, not one which will shut down part of our project if they misunderstand the credential policy. I am 90% sure, that most of them will misunderstand it; when I fix this now, it doesn't prevent someone else making the same mistake with another parallel process against some GCP API. The senior developer in me asks, what would be the best way to prevent the new guys from making this same mistake? Not using non-service accounts at all, ever?

**Comment by jba** [comment]
Are you using cloud.google.com/go/storage? Which calls are you making to that package?

As far as I can tell, we retry on 429 and 5xx, but nothing else. But more information would help.

**Comment by jba** [comment]
Never mind, I see what you're saying. You made many requests on the storage service using the wrong credentials.

It's hard to know how much "magic" to build in to a client. For example, we do automatic retries in some cases, as I mentioned in my previous comment. But for the most part, our client passes errors from the underlying service back to the caller, because the caller knows best how to handle them.

That rule should apply to 403s as much as any other error. In fact, your crawlers should probably back off or stop if they get more than a small fraction of errors from the storage service, regardless of the kind of error.

**Comment by nightlyone** [comment]
The issue boils down to "does it help, if I retry this operation?" and I guess for certain 403 errors, not related to the availability of credential files and account services, it makes sense to state: No retries won't help. Go has the interface with the Temporary() bool for answering that question. Now the api should simply implement this interface and return a proper value for Temporary. This would also leave retry frequency and delay in the hands of users while communicating clearly when retry is not beneficial due to persistent issues like wrong credentials.

**Comment by ThatGuyFromFinland** [comment]
I 100% agree, that the client user should be the one, who handles most errors. Less magic is always better, agreed.

I am doing streaming writes by the way, not writing entire objects at once.

Is there any sense to let the write attempts to the same object handler to go through to the API after it has failed? How likely is it, that the document will mystically appear (400), the authorization suddenly succeeds (401), the payment will be confirmed during the write attempt (402), which would actually lead to missing data in the file, the credentials will suddenly become valid (403), the object will mystically appear (404), someone would change the ACL and allow POSTing for the user during the write (405), which again would lead to a corrupt write and so on...

If it could be easily done, perhaps it would be better, that any further write attempts will not be sent to the API for the same object handler after receiving 4xx or 5xx error and perhaps some 3xx status codes?

What would be the reasonable use-cases where this kind of "keep on trying" mentality would be better, than the object locking itself, after which the user would be forced to solve the problem? Can any of the 4xx and 5xx errors be considered temporary or fixed without creating a new object or the human world doing something, which requires minutes of time? At least the timeouts and too-many-requests could get fixed, but keeping on polling the API would probably not be helpful in fixing those problems?

These errors will not fix themselves and the object itself will have to be able to do something differently in most cases; creating a new object is the solution most of the time. In which situations the user could halt the writing process and fix the problem and keep on using the same object handler? When writing parallel, which use cases that apply to the previous question, would not cause unstable resource allocation to the users system (queue building up)?

Actually I believe that if the object will not lock, the users are less likely to fix problems, such as timeouts and too-many-requests, because the API will kind of work; the 403 is the only situation, which leads to intrusion alerts other cases of the same mistake will just cause unnecessary load to the API.

**Comment by ThatGuyFromFinland** [comment]
@nightlyone Checked out the Temporary() bool. I agree; if the error code is not of temporary nature, the object handler should stop sending further write requests to the API. Defining explicitly, which status codes can be considered as temporary or not, might be a bit tricky in some cases. If object handler would be immutable (I don't know if it is, probably it is) making this distinction would be much easier.

**Comment by jba** [comment]
@ThatGuyFromFinland If you're talking about storage.Writer.Write, it already bypasses the network if there is an error from the Insert call. The Insert call error is set into w.err on https://github.com/GoogleCloudPlatform/google-cloud-go/blob/master/storage/writer.go#L99, and Write reads that on line 110 before it writes more data. However, I think there is a concurrency bug in that code, which should be fixed in https://code-review.googlesource.com/9952.

However, I don't think any of that really matters because as far as I can tell, the Insert RPC doesn't actually return until you close the writer. So there is no indication of an error until you write the whole object and close it. I tested that by creating a client with bad creds, then writing about a megabyte to an object, then closing the writer. I only saw the error when I closed. Are you observing different behavior?

It's also possible that I'm not understanding you correctly. Could you use specific symbols from the code instead of terms like "streaming writes" and "object handler"?

**Comment by ThatGuyFromFinland** [comment]
Sure!

I have a crawler processes which I launch as `go crawl()`, writeChan is a global variable where the crawlers dump whatever they find. There can be hundreds of concurrent crawlers running; the system is very stable.
```
for {
    select {
        ... other cases ...
        case payload = <- writeChan:
            persistor.Write(payload)
    }
}
```
Here is the Write method; this.writer is a storage.Writer...
```
func (this *Persistor) Write(payload []byte) {
    ... some irrelevant monitoring / performance code ...
    if _, err := this.writer.Write(payload); err != nil {
        fmt.Printf(err.Error() + "\n")
    }
    ... some irrelevant monitoring / performance code ...
}
```
... which is created here:
```
func (this *Persistor) setNewTarget() {
    ... some code to determine targetFileName etc. ...
    targetFile := this.bucket.Object(targetFileName)
    this.writer = targetFile.NewWriter(this.context)
}
```

With "object handler" I refer to targetFile; I only use one targetFile at time and grow it until it exceeds 40GB (json limit for BigQuery). With "streaming writes" I refer to the behavior, where storage.Writer.Write appends lines from my crawlers to the storage object and uploads the new lines to the API immediately.

I used to first write a file on local SDD and then upload it by running go.exec("gsutil"....), but the gsutil python scripts are too unstable to be used (allocates resources badly, which lead to crashing behavior very often if data was too dense; the uploads would become a bottleneck, since I would have to stop all other activity for them to safely execute). Besides, using SDD is much more expensive and slower, than streaming downloaded files to the upload stream straight from RAM.

The error happens, when I have wrong credentials; each storage.Writer.Write attempt will send data to the API resulting in 403, which looks like an intrusion alert. You don't have to close the storage.Writer before it starts sending in data to the API.

Now that I consider it, my proposal for fix is not sufficient. There can be many HTTP requests on their way to the server before the first 403 response will arrive.

Would it be too hard to make the storage.Writer check, that it has all the necessary rights to do the writing, before it starts writing? Of course that will create one extra API call per file written. However, such trade-off would be reasonable IMO considering, that writing many small files will anyways be slower than writing big ones, due to the cloud architecture. Another possibility would be to create different methods for streaming writes and single writes.

**Comment by jba** [comment]
> each storage.Writer.Write attempt will send data to the API resulting in 403

I was able to duplicate this (getting an error on Write) after I wrote about 8M. 

> Would it be too hard to make the storage.Writer check, that it has all the necessary rights to do the writing, before it starts writing? Of course that will create one extra API call per file written.

I don't think it makes sense for the client to do that check. It's a little too "magic," and it has the performance problem you mention.

However, I think it would make a lot of sense for your crawler to do that. On startup, it should write a 1-byte file to Cloud Storage.

**Comment by ThatGuyFromFinland** [comment]
That wouldn't work, because it has to be specific file and the ACL should check for write permissions on a specific file. Is it possible to try to write null?

It should be possible to check if the "object handler" has permission to write, before you actually write something. I think this kind of validity check would not be too "magic", since it makes sure, that the storage.Writer can actually do, what it promises to do: write the file.

**Comment by jba** [comment]
I confirmed that you can write a null []byte, then call Close, and you will get the error. You should double-check, because the way I am passing invalid creds gives me a 400, not a 403, so I'm not quite duplicating your case.

If we took your advice to its logical conclusion, then all our clients would check that they could perform an operation before actually trying to perform it. This would double the number of RPCs with no benefit, since it's always fine to just try the operation and handle the error. 

Your case is complicated by the facts that you have many concurrent writes, and there is some delay before the error comes back. In addition to writing null first, you might also consider using https://godoc.org/golang.org/x/sync/errgroup, which will cancel a context if any goroutine encounters an error.

**Comment by ThatGuyFromFinland** [comment]
With asynchronous appending writes it does not double the number of RPCs. The storage.Writer clearly is designed to be for asynchronous appending writes; otherwise having attributes like chunk size would not make sense.

It only adds one RPC per full stream of N writes. For example, I am writing 40GB files with over 12,000 writes per file.

You can't achieve write speed of over 400Mbit/s as easily by writing small files; besides that would not make sense for BigData anyway. I tried a different approach, where I was storing many files; that was incredibly slow and resource demanding due to the nature of HTTP protocol (for multiple files you can't keep the connection open, I guess).

I can fix my issue with null byte write, but I believe, that the API is a bit confusing. I would assume, that storage.Write is intended for writes, which are performed in (asynchronous) chunks; that would fit the Go philosophy. I think the developer should have a choice of using streaming writes or bulk writes, which would solve this problem elegantly. I think writing big files is a different issue, than writing small files.

To clarify: by using streaming writes, I can achieve faster crawling and persistance with much cheaper hardware. I don't have to use SSD, instead I read and write straight from the /dev/shm (tmpfs). This kind of streamlining is beneficial in many use cases as I avoid writing the file to a disk before starting the upload. By uploading on the go, I can write a system, where unexpected changes in data density (how much data I can crawl from a domain) are less complex to solve and will not cause fatal errors due to exceeded resource limits as easily.

**Comment by jba** [comment]
I think I understood everything you wrote, but I don't see how most of it is relevant. Let me try to summarize:

* The Google Cloud Storage service provides no clear way to check for write permissions without writing. That's not something we, the maintainers of the cloud.google.com/go/storage client, have any control over.
* A streaming write request to GCS doesn't return (with error or success) until around the first 8 meg have been written. Again, nothing we in this repo can control.
* To get the error early, then, someone has to write an empty file. It can be you, or it can be us.
* We won't do it, because it introduces extra hidden complexity into our client in the form of an extra RPC, which will double the number of RPCs for small (< 8M) files. (I think a lot of your last comment is arguing that streaming large files is a good thing, but just because it is doesn't mean our client should penalize writing small files.)
* Since you are writing such large files, the extra write won't impact your performance. And it is easy to code. You include it in an "open a GCS file for writing" function that all your developers call.

**Comment by ThatGuyFromFinland** [comment]
I agree, except that Google Cloud Storage will send intrusion alert within the first 30 write requests, which can fill up in much less than 8M.

Since the intrusion alerts are a feature of the Google Cloud Storage API, I believe the SDK should in some way recognize this. It would be much friendlier for the SDK user to not to learn about intrusion alerts existence the hard-way, where you have three days time to discover a hidden feature of the Google Cloud Storage API, which is not at all evident, when using the API.

I logged my errors and focused on right aspects, when the problem emerged. I was lucky in that sense.

I believe good SDK handles the critical errors of the API (like not accidently deleting the entire project), without the user having to be worried about the project being destroyed by an unexperienced developer.

As a senior developer I do not like SDK's, which can suspend my entire project from Google Cloud Platform, when one developer makes a simple mistake by launching the script, which uses the SDK, with wrong credentials. In order to enforce a secure way of operations, I would have to fork your SDK and maintain the fork myself.

Abstraction of the Google Cloud Platform features is the responsibility of Google Cloud Platform SDK. Writing large files with streaming inserts is not that special case. The intrusion alert mechanism clearly is a Google Cloud Platform feature.

I believe there should be two different methods:
-Bulk write, which closes the connection and the write handler straight after the write has been completed, but has limited amount of write requests ( = amount of requests allowed before intrusion alerts; total write size has to be less than chunk_size * intrusion alert limit), that will pass to the API.
-Streaming write, which will not close the connection before the write stream is closed; this method would first check for the credentials with null byte to avoid intrusion alerts.

I believe we now understand each other, and you can make the proper decision. In case I end up forking the project, I will post a link later on.

**Comment by adams-sarah** [comment]
Closing for now. Feel free to reopen if necessary.
